### PR TITLE
gcjob: teach GC job to respect protected timestamps for tables/indexes

### LINKED ISSUE #77156: gcjob: teach the GC job to respect protected timestamps on tables/indexes
The GC job currently relies on the `protectedts.Cache` to check if there are any protected timestamp records covering a span before allowing GC on the span. In the new protectedts multi-tenant subsystem, protected timestamp records no longer protect spans, but instead protect `ptpb.Targets` (table, database, tenant, cluster). This issue tracks the changes needed to fix this:

**For system tenants:**

To continue to work in a mixed version setting, GC jobs on the system tenant should continue to consult both the `protectedts.Cache` (for older span backed PTS records) and the new `ProtectedTimestampStateReader` that exports methods to check if a particular `Target` is protected.

**For secondary tenants:**

Secondary tenants have never written span-backed PTS records in older versions, and so they only need to consult the `ProtectedTimestampStateReader` to check if the table or index being gc'ed has a PTS record covering it.

Jira issue: CRDB-13439

### PR DESCRIPTION
**Pull request by arulajmani**
The GC job is responsible for issuing clear ranges for dropped tables
and indexes. Previously, we weren't consulting the new PTS subsystem
when deciding if a table or an index's range could be cleared. This
patch fixes that.

We check if a table/index span is protected after the GC TTL expires.
We do so by issuing an RPC to KV to fetch all span configurations and
system span configurations that apply over the spans. For secondary
tenants, this also includes any system span configurations that the
host tenant may have set over its ranges.

Lastly, it's worth noting that we make no attempt to coordinate with
concurrent protected timestamp writers. This is okay, given that
protected timestamps make no claims about expired data.

Closes https://github.com/cockroachdb/cockroach/issues/77156

Release note: None

### CODE REVIEW

File: pkg/sql/gcjob/refresh_statuses.go [hunk 1]
```diff
@@ -255,6 +267,85 @@ func getTableTTL(defTTL int32, zoneCfg *zonepb.ZoneConfig) int32 {
...
+			return false, err
+		}
+
+		// Collect all protected timestamps that apply to the given span; both by
+		// virtue of span configs and system span configs.
+		var protectedTimestamps []hlc.Timestamp
+		collectProtectedTimestamps := func(configs ...roachpb.SpanConfig) {
+			for _, config := range configs {
+				for _, protectionPolicy := range config.GCPolicy.ProtectionPolicies {
+					// We don't consider protected timestamps written by backups if the span
```
Comment (adityamaru) [inline pkg/sql/gcjob/refresh_statuses.go:315]: thank you 😋

File: pkg/ccl/testccl/sqlccl/tenant_gc_test.go [hunk 2]
```diff
@@ -130,6 +131,253 @@ func TestGCTenantRemovesSpanConfigs(t *testing.T) {
...
+			job, err := registry.LoadJob(ctx, sj.ID())
+			require.NoError(t, err)
+			require.Equal(t, jobs.StatusRunning, job.Status())
+
+			mu.Lock()
+			defer mu.Unlock()
+			if !mu.isSet {
+				return errors.Newf("waiting for isProtected status to be set")
+			}
+			require.Equal(t, true, mu.isProtected)
```
Comment (adityamaru) [inline pkg/ccl/testccl/sqlccl/tenant_gc_test.go:249]: nit: require.True if you want

File: pkg/ccl/testccl/sqlccl/tenant_gc_test.go [hunk 3]
```diff
@@ -130,6 +131,253 @@ func TestGCTenantRemovesSpanConfigs(t *testing.T) {
...
+		require.Equal(t, jobspb.SchemaChangeGCProgress_DELETED, progress.GetSchemaChangeGC().Tables[0].Status)
+
+		mu.Lock()
+		defer mu.Unlock()
+		require.Equal(t, false, mu.isProtected)
+	}
+
+	testutils.RunTrueAndFalse(t, "secondary-tenant", func(t *testing.T, forSecondaryTenant bool) {
+		resetTestingKnob()
+		if !forSecondaryTenant {
```
Comment (adityamaru) [inline pkg/ccl/testccl/sqlccl/tenant_gc_test.go:271]: leftover from some debugging?

File: pkg/ccl/testccl/sqlccl/tenant_gc_test.go [hunk 4]
```diff
@@ -130,6 +131,253 @@ func TestGCTenantRemovesSpanConfigs(t *testing.T) {
...
+		)
+
+		// We do something similar for a PTS set by the host over the tenant's
+		// keyspace.
+		hostOnTenant, err := spanconfig.MakeTenantKeyspaceTarget(roachpb.SystemTenantID, tenantID)
+		require.NoError(t, err)
+		r2, err := spanconfig.MakeRecord(
+			spanconfig.MakeTargetFromSystemTarget(hostOnTenant),
+			makeSystemSpanConfig(6, 28, 35),
+		)
```
Comment (adityamaru) [inline pkg/ccl/testccl/sqlccl/tenant_gc_test.go:303]: missing require.NoError()?

File: pkg/ccl/testccl/sqlccl/tenant_gc_test.go [hunk 5]
```diff
@@ -130,6 +131,253 @@ func TestGCTenantRemovesSpanConfigs(t *testing.T) {
...
+		resetTestingKnob()
+
+		if !forSecondaryTenant {
+			ensureGCSucceeded(t, registry, sj)
+			return
+		}
+
+		// Next, we'll remove the host set system span config on our secondary
+		// tenant.
+		r1, err = spanconfig.MakeRecord(
```
Comment (adityamaru) [inline pkg/ccl/testccl/sqlccl/tenant_gc_test.go:359]: should this be r2 with 28, 35. Though I see we're deleting the whole target below so maybe this is just not needed?

File: pkg/ccl/testccl/sqlccl/tenant_gc_test.go [hunk 6]
```diff
@@ -130,6 +131,253 @@ func TestGCTenantRemovesSpanConfigs(t *testing.T) {
...
+		sj, err := jobs.TestingCreateAndStartJob(
+			ctx, registry, execCfg.DB, makeGCJobRecord(10, descpb.ID(tableID)),
+		)
+		require.NoError(t, err)
+
+		ensureGCBlockedByPTS(t, registry, sj)
+
+		// Update PTS state that applies to the entire keyspace -- we only
+		// remove the timestamp before the drop time. We should expect GC to
+		// succeed on if we're not testing for the system tenant at this point.
```
Comment (adityamaru) [inline pkg/ccl/testccl/sqlccl/tenant_gc_test.go:341]: This seems a little at odds with the outline at the top of the test. Isn't the expectation that removing the entire keyspace protection policy below the drop time, affects both the system and secondary tenants? i.e. allows GC on both? I think thats what we are checking below bu the last line of this comment is a little misleading.

File: pkg/ccl/testccl/sqlccl/tenant_gc_test.go [hunk 7]
```diff
@@ -130,6 +131,253 @@ func TestGCTenantRemovesSpanConfigs(t *testing.T) {
...
+		err = tsKVAccessor.UpdateSpanConfigRecords(ctx, nil /*toDelete */, []spanconfig.Record{r1})
+		require.NoError(t, err)
+
+		resetTestingKnob()
+
+		if !forSecondaryTenant {
+			ensureGCSucceeded(t, registry, sj)
+			return
+		}
+
```
Comment (adityamaru) [inline pkg/ccl/testccl/sqlccl/tenant_gc_test.go:356]: can we check that GC has still not succeeded if we are on a secondary tenant?

**Review by adityamaru — CHANGES_REQUESTED** [review]
Nice! Just some minor comments, will stamp after another round.

**Review by arulajmani — COMMENTED** [review]
Addressed all your comments, thanks for the quick review! 

Looks like we're failing `TestFullClusterBackup` in CI. This seems to be because we've disabled the span config infra for that test (see: https://github.com/cockroachdb/cockroach/issues/75060), which means the KVAccessor we use to determine if tables can be GC-ed perennially errors out. As a result, we never GC, and fail the expectation that there are no running GC jobs in the test. I can shove in a testing knob to skip checking the new subsystem in the GC job. Are you on board with that? 

Separately, @adityamaru, is this test something we want to de-flake and run with the new subsystem before the release goes out? Is this something Bulk is tracking? 

> Dismissed @adityamaru from 3 discussions.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78475#-:-Mz0cwnPEy9FzHWGlzKp:b-cjrlog)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, and @nvanbenschoten)

___
*[pkg/roachpb/span_config.proto, line 342 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78475#-MyzOaN76q-MQ341sGno-r1-342:-Mz0VtD_4JsuGmiAGuxJ:bt0qmvy) ([raw file](https://github.com/cockroachdb/cockroach/blob/f9c74e45c53f618ced6dab48b909b271ec414aa3/pkg/roachpb/span_config.proto#L342)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

nit: Are we missing a word here? TenantID identifies the tenant...
</blockquote></details>

Few words actually; fixed.
___
*[pkg/sql/gcjob/refresh_statuses.go, line 315 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/78475#-MyzOaN8ayGLi7ounrfH-r2-315:-Mz0_6Kj2YSCCTnTLyJM:bb6wwgn) ([raw file](https://github.com/cockroachdb/cockroach/blob/16057beb5208d82484ce302189b66681e2602160/pkg/sql/gcjob/refresh_statuses.go#L315)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

thank you 😋 
</blockquote></details>

Haha
___
*[pkg/ccl/testccl/sqlccl/tenant_gc_test.go, line 271 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/78475#-MyzOaN8ayGLi7ounrfE-r2-271:-Mz0WrJ11EjIRs9OltZu:bo6ap9z) ([raw file](https://github.com/cockroachdb/cockroach/blob/16057beb5208d82484ce302189b66681e2602160/pkg/ccl/testccl/sqlccl/tenant_gc_test.go#L271)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

leftover from some debugging? 
</blockquote></details>

Indeed!
___
*[pkg/ccl/testccl/sqlccl/tenant_gc_test.go, line 303 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/78475#-MyzOaN8ayGLi7ounrfE-r2-303:-Mz0Y1zH5K_lJB8nhe3b:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/16057beb5208d82484ce302189b66681e2602160/pkg/ccl/testccl/sqlccl/tenant_gc_test.go#L303)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

missing require.NoError()?
</blockquote></details>

Done.
___
*[pkg/ccl/testccl/sqlccl/tenant_gc_test.go, line 341 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/78475#-MyzOaN8ayGLi7ounrfE-r2-341:-Mz0YvMA3RV4zoSqX7Zf:b-u26nsp) ([raw file](https://github.com/cockroachdb/cockroach/blob/16057beb5208d82484ce302189b66681e2602160/pkg/ccl/testccl/sqlccl/tenant_gc_test.go#L341)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

This seems a little at odds with the outline at the top of the test. Isn't the expectation that removing the entire keyspace protection policy below the drop time, affects both the system and secondary tenants? i.e. allows GC on both? I think thats what we are checking below bu the last line of this comment is a little misleading.
</blockquote></details>

Oops, wrongly worded comment. It should've said "we should expect GC to succeed if we're testing for the system tenant at this point", not the other way around. Fixed. 
___
*[pkg/ccl/testccl/sqlccl/tenant_gc_test.go, line 356 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/78475#-MyzOaN8ayGLi7ounrfE-r2-356:-Mz0ZcKDA6O8ybVuR0k1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/16057beb5208d82484ce302189b66681e2602160/pkg/ccl/testccl/sqlccl/tenant_gc_test.go#L356)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

can we check that GC has still not succeeded if we are on a secondary tenant?
</blockquote></details>

Done.
___
*[pkg/ccl/testccl/sqlccl/tenant_gc_test.go, line 359 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/78475#-MyzOaN8ayGLi7ounrfE-r2-359:-Mz0Zi-I33Isr8GMvnuQ:b-ppanvl) ([raw file](https://github.com/cockroachdb/cockroach/blob/16057beb5208d82484ce302189b66681e2602160/pkg/ccl/testccl/sqlccl/tenant_gc_test.go#L359)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

should this be r2 with 28, 35. Though I see we're deleting the whole target below so maybe this is just not needed?
</blockquote></details>

I think I was just meant to delete this. I initially had grand plans for this integration test, but as you can see, it's super heavy weight.

**Comment by adityamaru** [comment]
> Is this something Bulk is tracking?

My impression was that this would be picked up when addressing the fallout from enabling the spanconfig infra, so afaik Bulk didn't really think we owned this. I had filed https://github.com/cockroachdb/cockroach/issues/70173 a while ago and I believe it still needs to be addressed. The two may be related.

> Are you on board with that?
Onboard with it to unblock this yep, maybe a TODO saying that we should get rid of the knob at some point.

**Review by nvb — COMMENTED** [review]
> Reviewed 8 of 15 files at r1, 18 of 18 files at r3, 10 of 10 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78475)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, and @arulajmani)

___
*[pkg/roachpb/span_config.proto, line 336 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/78475#-MzCAnlg2Wqum31iQ-8k:-MzCAnlg2Wqum31iQ-8l:bv3miwi) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e1b0ee05350294b303e62d82e5089c3ddaa2be4/pkg/roachpb/span_config.proto#L336)):*
> ```protobuf
> // configs that apply to a given tenant. This is only ever meaningful in the
> // context of secondary tenants as it allows them a view into system span config
> // state that applies over their ranges that has been set by the system tenant.
> ```

This might be a good place to remind readers what "apply" means in this context.
___
*[pkg/sql/gcjob/refresh_statuses.go, line 177 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/78475#-MzCBckJBiwJWd00VVoX:-MzCBckJBiwJWd00VVoY:bcwj6y4) ([raw file](https://github.com/cockroachdb/cockroach/blob/387018472871bc3016b179f1472ccb49f35b2e9e/pkg/sql/gcjob/refresh_statuses.go#L177)):*
> ```Go
> 		if err != nil {
> 			log.Infof(ctx, "error checking protection status %v", err)
> 			// Retry again later?
> ```

Was this a comment that you intended to address in this PR? Also, should ew use a more severe `log` level, like Warning or Error?
___
*[pkg/sql/gcjob/gc_protected_timestamp_test.go, line 72 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/78475#-MzCCOZx0EmojMd0uQ8w:-MzCCOZx0EmojMd0uQ8x:brfrop8) ([raw file](https://github.com/cockroachdb/cockroach/blob/387018472871bc3016b179f1472ccb49f35b2e9e/pkg/sql/gcjob/gc_protected_timestamp_test.go#L72)):*
> ```Go
> 		setup         func(t *testing.T, kvAccessor *manualKVAccessor, cache *manualCache)
> 		isProtected   bool
> 		droppedAtTime int64
> ```

nit: the ordering and naming of fields in this struct doesn't make it very clear what's the input and what's the expected output. Consider switching around `droppedAtTime` and `isProtected` and renaming the latter to `expectIsProtected`.

**Review by arulajmani — COMMENTED** [review]
Turns out I hadn't switched out the `deprecatedIsProtected` codepath for indexes -- I fixed that and added an index variant to the integration test I had for tables. PTAL! 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78475#-:-MzDfHY15csjM9U4pK63:b-cssx57)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, and @nvanbenschoten)

___
*[pkg/roachpb/span_config.proto, line 336 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/78475#-MzCAnlg2Wqum31iQ-8k:-MzDEnsaAGjmqGZrpNKR:b-qy7sib) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e1b0ee05350294b303e62d82e5089c3ddaa2be4/pkg/roachpb/span_config.proto#L336)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This might be a good place to remind readers what "apply" means in this context.
</blockquote></details>

Added. 
___
*[pkg/sql/gcjob/refresh_statuses.go, line 177 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/78475#-MzCBckJBiwJWd00VVoX:-MzDdm5G4k8B8ZgTQd4N:bg57o3k) ([raw file](https://github.com/cockroachdb/cockroach/blob/387018472871bc3016b179f1472ccb49f35b2e9e/pkg/sql/gcjob/refresh_statuses.go#L177)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Was this a comment that you intended to address in this PR? Also, should ew use a more severe `log` level, like Warning or Error?
</blockquote></details>

Nice catch! There's not much we need to do here, other than return something sane, as the GC job will automatically be retried above in the call stack. I added a comment explaining what's happening and changed the log level to be Error. 
___
*[pkg/sql/gcjob/gc_protected_timestamp_test.go, line 72 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/78475#-MzCCOZx0EmojMd0uQ8w:-MzDFXk-A-eZ3CR0t2hO:bs7wumx) ([raw file](https://github.com/cockroachdb/cockroach/blob/387018472871bc3016b179f1472ccb49f35b2e9e/pkg/sql/gcjob/gc_protected_timestamp_test.go#L72)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: the ordering and naming of fields in this struct doesn't make it very clear what's the input and what's the expected output. Consider switching around `droppedAtTime` and `isProtected` and renaming the latter to `expectIsProtected`.
</blockquote></details>

Done.

**Review by adityamaru — APPROVED** [review]

**Comment by arulajmani** [comment]
TFTRs! 

bors r+

**Comment by arulajmani** [comment]
Flaked, we go again. 

bors r+

**Pull request merged**